File: ecos-records/src/main/java/ru/citeck/ecos/records3/record/op/atts/proc/AbstractAttProcessor.java
package ru.citeck.ecos.records3.record.op.atts.proc;

import org.jetbrains.annotations.NotNull;
import ru.citeck.ecos.commons.data.DataValue;
import ru.citeck.ecos.commons.data.ObjectData;

import java.util.List;

public abstract class AbstractAttProcessor<T> implements AttProcessor {

    private boolean allowNull;

    public AbstractAttProcessor(boolean allowNull) {
        this.allowNull = allowNull;
    }

    public AbstractAttProcessor() {
    }

    @NotNull
    public Object process(@NotNull ObjectData meta, @NotNull DataValue value, @NotNull List<DataValue> arguments) {

        if (!allowNull && value.isNull()) {
            return value;
        }

        T args = parseArgs(arguments);

        if (value.isArray()) {
            DataValue res = DataValue.createArr();
            value.forEach(v -> res.add(processOne(meta, v, args)));
            return res;
        } else {
            return processOne(meta, value, args);
        }
    }

    protected abstract Object processOne(@NotNull ObjectData meta, @NotNull DataValue value, @NotNull T arguments);

    protected abstract T parseArgs(@NotNull List<DataValue> arguments);
}

File: ecos-records/src/main/java/ru/citeck/ecos/records3/record/op/atts/service/proc/AbstractAttProcessor.kt
package ru.citeck.ecos.records3.record.op.atts.service.proc

import ru.citeck.ecos.commons.data.DataValue
import ru.citeck.ecos.commons.data.ObjectData

abstract class AbstractAttProcessor<T> : AttProcessor {

    private var allowNull = false

    constructor()

    constructor(allowNull: Boolean) {
        this.allowNull = allowNull
    }

    override fun process(attributes: ObjectData, value: DataValue, args: List<DataValue>): Any? {
        if (!allowNull && value.isNull()) {
            return value
        }
        val arguments = parseArgs(args)
        return if (value.isArray()) {
            val res: DataValue = DataValue.createArr()
            value.forEach { res.add(processOne(attributes, it, arguments)) }
            res
        } else {
            processOne(attributes, value, arguments)
        }
    }

    protected abstract fun processOne(attributes: ObjectData, value: DataValue, args: T): Any?

    protected abstract fun parseArgs(args: List<DataValue>): T
}
